### DISCUSSION TITLE
Error : subscript out of bounds

### DISCUSSION THREAD

**Discussion by tu849035** [teamalgoritma/community#404] [Q&A] [answered]
Halo.

Saat saya akan mengambil data hasil PCA sebanyak PC yang dibutuhkan, muncul error seperti tampilan di bawah.

![image](https://user-images.githubusercontent.com/107348211/197000966-5db8a012-fcc9-4ddd-8be5-b4a2f2bae4bd.png)

Mohon bantuan dan solusinya.

Terima Kasih.

**Comment by jafarhf1** [comment]
Dear @tu849035 ,
Berdasarkan tulisan error dari pertanyaan yang diajukan, maka kemungkinan berasal dari bagian kode `whiskey_pca$eig[,1:8]`.  Di situ tertulis bahwa subset yang bapak lakukan tidak sesuai dengan jumlah dari data yang ingin disubset. Bapak ingin melakukan subset sebanyak 8 kolom sedangkan data `whiskey_pca$eig` kemungkinan hanya memiliki kolom kurang dari 8. Mungkin bapak bisa coba dicek kembali untuk data `whiskey_pca$eig` nya berapa jumlah kolomnya.

Demikian yang dapat kami sampaikan, semoga membantu.

**Comment by tu849035** [comment]
Dear mas @jafarhf1 ,

Kalau saya cek summarynya seperti tampilan di bawah mas.

![image](https://user-images.githubusercontent.com/107348211/197122867-aec5367b-c654-4ae9-8925-bd614c203d7f.png)

Ada 12 dimensinya. Dan saya cek berdasarkan eigenvalue nya pun ada 12 comp seperti tampilan di bawah.

![image](https://user-images.githubusercontent.com/107348211/197122988-b4a74283-bf11-4d2d-b7c4-62c7c814ff8f.png)

Karena saya ingin ambil variance nya di 90%, makanya saya mau ambil yang comp 1 - 8 saja.
Tapi, kenapa jadi [Error : subscript out of bounds] ya?

Mohon bantuan & solusinya.

Terima Kasih.

**Reply by jafarhf1** [reply]
Dear  @tu849035 
Berdasarkan data `whiskey_pca$eig` yang bapak perlihatkan, disitu terdiri dari 12 baris dan 3 kolom. Oleh karena itu, jika bapak ingin mengambil comp 1-8 saja, maka bapak melakukan subsetting berdasarkan baris seperti ini  `whiskey_pca$eig[1:8,]`

Semoga membantu pak.

**Reply by tu849035** [reply]
Dear mas @jafarhf1 ,

Kalau mengikuti saran das Mas nya, bentuk data framenya jadi berbeda dari yang saya pejalari sebelumnya di kelas ya mas.
![image](https://user-images.githubusercontent.com/107348211/197125343-bec30217-fe43-454c-a413-8ec9c9cf4199.png)

Sedangan saat dikelas, tampilan dataframenya seperti ini.
![image](https://user-images.githubusercontent.com/107348211/197125486-1b22a78b-aeea-4f44-8673-6cad9e799777.png)

Kalau saya menggunakan dataframe `whiskey_pca$eig[1:8,]`, apakah masih bisa digunakan utk keperluan analisa ke depannya sama seperti menggunakan dataframe yang `loan_pca$ind$coord[, 1:5])`? seperti pada in-class.

Dan sebaiknya utk Dimensionality Reduction, saat pembuatan dataframe yang baru saya menggunakan yang mana ya mas?

Mohon bantuan dan solusinya.

Terima Kasih.

**Comment by ranydwi** [comment, marked as answer]
Halo, selamat siang Pak @tu849035.

Kalau dari yang saya lihat, seharusnya dari hasil summary Bapak yang mana memiliki 12 dimensi, kemudian Bapak ingin lakukan reduksi sebanyak 8 dimensi saja. Code yang Bapak tuliskan sudah benar, yakni `whiskey_pca$ind$coord[, 1:8]` . Tidak ada issue untuk itu Pak.

Baik Pak, saya coba jelaskan dari error yang Bapak temui.
1. Error yang dihasilkan itu berasal dari code Bapak yang ini `whiskey_pca$eig[ ,1:8]`. Mengapa? Karena sesuai jawaban Bapak, output ketika kita menjalankan `whiskey_pca$eig` memiliki 12 baris dan 3 kolom. Tidak tepat ketika kita malah mau melakukan subseting kolom 1-8, karena kolomnya yg ada saja hanya 3. Sehingga yang tepat sesuai dengan arahan Mas Jafar yaitu `whiskey_pca$eig[1:8, ]`.
2. Ketika Bapak inginnya mengambil data hasil PCA sebanyak PC yang dibutuhkan, maka untuk melakukan dimentionality reduction adalah menggunakan **`whiskey_pca$ind$coord[, 1:8]` yaa Pak. Karena intuisi dari `$ind$coord` bermakna mengambil data hasil PCA seperti `$x` apabila Bapak melakukan analisis PCA pada base**. 
3. Untuk pertanyaan Bapak yang ini "Kalau saya menggunakan dataframe whiskey_pca$eig[1:8,], apakah masih bisa digunakan utk keperluan analisa ke depannya sama seperti menggunakan dataframe yang loan_pca$ind$coord[, 1:5])? seperti pada in-class.", jawabannya adalah: **tidak bisa Pak, karena makna dari `$eig` hanyalah informasi mengenai banyaknya informasi yang berhasil ditangkap untuk masing-masing dimensi**. Silahkan untuk dimaknai yaa Pak:)

Sehingga kesimpulan error Bapak, yaitu karena bapak salah melakukan subseting ketika ingin mengambil nilai `whisky_pca$eig[ , 1:8]`. Sementara berdasarkan tujuan Bapak silahkan saja langsung lakukan subseting untuk `whiskey_pca$ind$coord[, 1:8]` dan tidak perlu melakukan subseting untuk `whisky_pca$eig[ , 1:8]` ya Pak Tubagus.

Demikian yang dapat kami sampaikan, semoga membantu.

**Reply by tu849035** [reply]
Dear Mbak @ranydwi ,

Baik mbak saya sudah paham.
Tapi, ini saya menggunakan `whiskey_pca$ind$coord[, 1:8]` masih error mbak.
![image](https://user-images.githubusercontent.com/107348211/197142509-f48861f5-1c98-449a-9b99-1b4893d2ca10.png)

Dan setelah saya cek `whiskey_pca$ind$coord` ini hanya memiliki 5 dimensi. Itu kenapa ya mbak? Padahal `whisky_pca$eig` ini ada 12 dimensi.
Sepertinya karena dimensinya berbeda makanya `whiskey_pca$ind$coord[, 1:8]` masih error.
![image](https://user-images.githubusercontent.com/107348211/197143011-94962348-964d-4561-bbb1-219d67b2420f.png)

Mohon bantuan dan solusinya.

Terima Kasih.

**Reply by ranydwi** [reply]
Dear Pak @tu849035,

Berdasarkan konsep PCA, seharusnya hal ini tidak mungkin terjadi Pak Tubagus. Karena nilai `whisky_pca$ind$coord` seharusnya sesuai dengan banyaknya variabel yang ada pada data (dalam hal Bapak 12). Saya curiga, apakah sebelumnya Bapak sudah melakukan running untuk subseting sebanyak 5? Boleh coba untuk di running kembali object PCA Bapak. Kemudian coba untuk dikeluarkan kembali `whisky_pca$ind$coord` yaa Pak.

Semoga membantu, semangat!
